lib/spack/spack/solver/asp.py

```diff
@@ -2000,7 +1999,10 @@ def define_ad_hoc_versions_from_specs(
             # best possible, so they're guaranteed to be used preferentially.
             version = s.versions.concrete
 
-            if version is None or any(v == version for v in self.possible_versions[s.name]):
+            if version is None or any(
+                (v == version and not isinstance(version, vn.GitVersion))
+                for v in self.possible_versions[s.name]
+            ):
```

Comment:
This does extra comparisons relative to:

```suggestion
            if version is None or (not isinstance(version, vn.GitVersion) and any(
                v == version
                for v in self.possible_versions[s.name]
            )):
```

Reply:
Yeah good call. Out of curiosity, why do we even have that loop? Are there other cases besides `GitRefVersions` where a user can supply an arbitrary version on the CLI and we expect it to concretize?

Also I wonder if we should not be allowing any arbitrary version at the CLI, but rather require the RHS of the `GItRefVersion` to be known to spack already?